lib/charms/mlops_libs/v0/k8s_service_info.py

```diff
@@ -0,0 +1,297 @@
+        refresh_event: (list, optional): list of BoundEvents that this manager should handle.  Use this to update
+                       the data sent on this relation on demand.
+
+    Attributes:
+        provider_charm (CharmBase): variable for storing the provider application
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(
+        self,
+        provider_charm: CharmBase,
+        relation_name: str,
+        svc_name: str,
+        svc_port: str,
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
+    ):
+        super().__init__(provider_charm, relation_name)
+        self.provider_charm = provider_charm
+        self.relation_name = relation_name
+        self._requirer_wrapper = KubernetesServiceInfoProviderWrapper(
+            self.provider_charm, self.relation_name
+        )
+        self._svc_name = svc_name
+        self._svc_port = svc_port
+
+        self.framework.observe(self.provider_charm.on.leader_elected, self._send_k8s_svc_info)
+
+        self.framework.observe(
+            self.provider_charm.on[self.relation_name].relation_crated, self._send_k8s_svc_info
+        )
+
+        if refresh_event:
+            if not isinstance(refresh_event, (tuple, list)):
+                refresh_event = [refresh_event]
+            for evt in refresh_event:
+                self.framework.observe(evt, self._send_k8s_svc_info)
+
+        def _send_k8s_svc_info(self, _):
+            """Serve as an event handler for sending the Kubernetes Service information."""
+            self._requirer_wrapper.send_k8s_svc_info_relation_data(self._svc_name, self._svc_port)
+
+
+class KubernetesServiceInfoProviderWrapper(Object):
+    """Base class that represents a provider relation end.
+
+    Args:
+        provider_charm (CharmBase): the provider application
+        relation_name (str, optional): the name of the relation
+
+    Attributes:
+        provider_charm (CharmBase): variable for storing the provider application
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(self, provider_charm: CharmBase, relation_name: str):
+        super().__init__(provider_charm, relation_name)
+        self.provider_charm = provider_charm
+        self.relation_name = relation_name
+
+    def send_k8s_svc_info_relation_data(
+        self,
+        svc_name: str,
+        svc_port: str,
+    ) -> None:
+        """Update the relation data bag with data from a Kubernetes Service.
+
+        This method will complete successfully even if there are no related applications.
+
+        Args:
+            svc_name (str): the name of the Kubernetes Service the provider knows about
+            svc_port (str): the port number of the Kubernetes Service the provider knows about
+        """
+        # Update the relation data bag with a Kubernetes Service information
+        relations = self.model.relations[self.relation_name]
```

Comment:
Should this be
```suggestion
        relations = self.provider_charm.model.relations[self.relation_name]
```

?

Reply:
If I'm right about this item above, I think that suggests a gap in test coverage as it should have caused an error.  Is there a test we could add that would have caught this?

Reply:
I'll update the test accordingly to avoid this.